File: 05-adversarial/src/Ledger.java
// 05-adversarial/src/Ledger.java
// A minimal points ledger. Accounts hold points, and transfer moves points between accounts.
// Plain Java, no framework.
import java.util.HashMap;
import java.util.Map;

public class Ledger {
    private final Map<String, Integer> balances = new HashMap<>();

    public Ledger(Map<String, Integer> initial) {
        balances.putAll(initial);
    }

    public int balanceOf(String account) {
        return balances.getOrDefault(account, 0);
    }

    public void transfer(String from, String to, int amount) {
        int available = balanceOf(from);
        if (available < amount) {
            throw new IllegalStateException("insufficient balance");
        }
        balances.put(from, available - amount);
        balances.put(to, balanceOf(to) + amount);
    }
}

File: 05-adversarial/src/Ledger.cs
// 05-adversarial/src/Ledger.cs
// A minimal points ledger. Accounts hold points, and transfer moves points between accounts.
// Plain C#, no framework.
using System;
using System.Collections.Generic;

public class Ledger
{
    private readonly Dictionary<string, int> _balances = new();

    public Ledger(IDictionary<string, int> initial)
    {
        foreach (var entry in initial)
        {
            _balances[entry.Key] = entry.Value;
        }
    }

    public int BalanceOf(string account)
    {
        return _balances.TryGetValue(account, out var balance) ? balance : 0;
    }

    public void Transfer(string from, string to, int amount)
    {
        int available = BalanceOf(from);
        if (available < amount)
        {
            throw new InvalidOperationException("insufficient balance");
        }
        _balances[from] = available - amount;
        _balances[to] = BalanceOf(to) + amount;
    }
}
